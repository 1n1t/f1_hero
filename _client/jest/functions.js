/* eslint-disable import/prefer-default-export */
export function sel(id) {
	return `[data-test="${id}"]`;
}
