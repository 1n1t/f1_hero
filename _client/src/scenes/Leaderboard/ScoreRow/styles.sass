.ScoreRow
  display: flex
  padding: 0 30px
  &:not(:last-child)
    margin-bottom: 20px

  .username
    flex-grow: 1
    margin-left: 20px
