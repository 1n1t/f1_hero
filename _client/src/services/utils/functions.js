/* eslint-disable import/prefer-default-export */
export const isObjectEmpty = obj =>
	Object.keys(obj).length === 0 && obj.constructor === Object;

export const getNodePaddings = node => {
	const nodeStyle = window.getComputedStyle(node);
	return (
		parseInt(nodeStyle.getPropertyValue('padding-right'), 10) +
		parseInt(nodeStyle.getPropertyValue('padding-left'), 10)
	);
};
