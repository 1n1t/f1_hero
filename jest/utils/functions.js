const mongoose = require('mongoose');
const transformProps = require('transform-props');

function clearCollection(model) {
	return mongoose.model(model).remove();
}

function clearAllCollections() {
	Promise.all([
		mongoose.model('race').remove(),
		mongoose.model('user').remove(),
		mongoose.model('pick').remove(),
		mongoose.model('driver').remove()
	]).then(() => null);
}

function castToString(arg) {
	return String(arg);
}

function docToObj(docObj) {
	return transformProps(docObj, castToString, '_id');
}

module.exports = {
	clearCollection,
	clearAllCollections,
	docToObj
};
