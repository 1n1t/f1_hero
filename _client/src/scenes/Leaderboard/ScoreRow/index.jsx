import React, { Component } from 'react';
import PropTypes from 'prop-types';

import './styles.scss';

class ScoreRow extends Component {
	shouldComponentUpdate() {
		return false;
	}

	render() {
		const { index, user, total } = this.props;

		return (
			<div className="ScoreRow">
				<div>{index + 1}.</div>
				<div className="username">{user}</div>
				<div>{total}pt</div>
			</div>
		);
	}
}

ScoreRow.propTypes = {
	user: PropTypes.string.isRequired,
	total: PropTypes.number.isRequired,
	index: PropTypes.number.isRequired
};

export default ScoreRow;
