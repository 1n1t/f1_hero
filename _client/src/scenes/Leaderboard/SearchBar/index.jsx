import React from 'react';
import PropTypes from 'prop-types';

import './styles.scss';

const SearchBar = ({ searchTerm, onChange }) => (
	<div className="SearchBar">
		<input
			type="text"
			placeholder="Search..."
			onChange={onChange}
			value={searchTerm}
			data-test="searchInput"
		/>
		<i className="fa fa-search" aria-hidden="true" />
	</div>
);

SearchBar.propTypes = {
	searchTerm: PropTypes.string.isRequired,
	onChange: PropTypes.func.isRequired
};

export default SearchBar;
