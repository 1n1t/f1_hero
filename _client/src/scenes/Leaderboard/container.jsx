/* eslint-disable react/prop-types */
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { getUsersScores } from '../../data/leaderboard/actions';
import { updatePageHeading } from '../../services/page/actions';
import Leaderboard from './index';

const mapStateToProps = state => ({ scores: state.data.leaderboard.scores });

class LeaderboardContainer extends Component {
	componentDidMount() {
		this.props.updatePageHeading('Leaderboard');
		this.props.getUsersScores();
	}

	render() {
		return <Leaderboard scores={this.props.scores} />;
	}
}

export default connect(mapStateToProps, { getUsersScores, updatePageHeading })(
	LeaderboardContainer
);
