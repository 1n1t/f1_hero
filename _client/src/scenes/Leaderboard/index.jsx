import React, { Component } from 'react';
import PropTypes from 'prop-types';

import ScoreRow from './ScoreRow';
import SearchBar from './SearchBar';
import './styles.scss';

class Leaderboard extends Component {
	state = { searchTerm: '' };

	handleSearchTermChange = event => {
		const newState = {
			...this.state,
			searchTerm: event.target.value
		};
		this.setState(newState);
	};

	renderResultsList() {
		return this.props.scores
			.filter(
				score =>
					score.user
						.toLowerCase()
						.indexOf(this.state.searchTerm.toLowerCase()) >= 0
			)
			.map((score, index) => (
				<ScoreRow {...score} index={index} key={score.user} />
			));
	}

	render() {
		return (
			<div className="Leaderboard">
				<SearchBar
					onChange={this.handleSearchTermChange}
					searchTerm={this.state.searchTerm}
				/>
				<div className="results">{this.renderResultsList()}</div>
			</div>
		);
	}
}

Leaderboard.defaultProps = {
	scores: []
};

Leaderboard.propTypes = {
	scores: PropTypes.arrayOf(
		PropTypes.shape({
			user: PropTypes.string,
			total: PropTypes.number
		})
	)
};

export default Leaderboard;
