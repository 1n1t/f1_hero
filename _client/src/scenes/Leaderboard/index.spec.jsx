import r from 'Root/jest/utils/renderer';
import leaderboardSample from 'JestClient/samples/leaderboard';
import { sel } from 'JestClient/functions';

import Leaderboard from './index';
import ScoreRow from './ScoreRow';

const defaultProps = {
	scores: leaderboardSample.scores
};

r.init(Leaderboard, defaultProps);

it('renders correctly', () => {
	const wrapper = r.render(null, true);
	expect(wrapper).toMatchSnapshot();
});

it('renders correct amount of users scores', () => {
	const wrapper = r.render(null, true);
	const rows = wrapper.find(ScoreRow);

	expect(rows).toHaveLength(defaultProps.scores.length);
});

it('does not renders scores if array is empty', () => {
	const wrapper = r.render({ scores: [] }, true);
	const rows = wrapper.find(ScoreRow);

	expect(rows).toHaveLength(0);
});

it('filters score list according to provided search term', () => {
	const searchTerm = 'Jo';
	const wrapper = r.render(null, false);

	const input = wrapper.find(sel('searchInput'));
	input.simulate('change', { target: { value: searchTerm } });

	const rows = wrapper.update().find(ScoreRow);
	const newScores = defaultProps.scores.filter(
		({ user }) => user.toLowerCase().indexOf(searchTerm.toLowerCase()) >= 0
	);

	expect(rows).toHaveLength(newScores.length);
});
