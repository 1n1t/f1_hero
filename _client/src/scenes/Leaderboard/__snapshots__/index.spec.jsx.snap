// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="Leaderboard"
>
  <SearchBar
    onChange={[Function]}
    searchTerm=""
  />
  <div
    className="results"
  >
    <ScoreRow
      index={0}
      key="John Doe"
      total={123}
      user="John Doe"
    />
    <ScoreRow
      index={1}
      key="Mary Doe"
      total={95}
      user="Mary Doe"
    />
    <ScoreRow
      index={2}
      key="Josh Simpson"
      total={12}
      user="Josh Simpson"
    />
  </div>
</div>
`;
